[Experimental] Add block text color support to interactivity filter blocks

samueljseay opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR adds support for the block "text" color setting when displaying the experimental interactivity API powered collection filter blocks. The only block not currently supported is the price filter. By default adding text color colors the bar of the range filter UX control and I have a question whether the input text should be colored. I would like to do that one as a separate follow up to keep PRs small.

To do this:

1. We allow the setting in block.json
2. On the editor side the checkbox list mode already works automatically, but for dropdown we just need to display the placeholder text as the correct color to simulate. To do this we use `@emotion/styled` directly on FormTokenField component
3. On PHP side we use the existing style attribute utils to get the colors from block props, we pass these to the `Dropdown` component. To target the placeholder due to it being a pseudo element we inject a small stylesheet when rendering. For other text in the dropdown we just apply an inline style targeting `color`.

Partially resolves #42056 (just note there I have the question about text color support for price range slider)
<img width="689" alt="Screenshot 2023-12-22 at 2 01 16 PM" src="https://github.com/woocommerce/woocommerce/assets/1281828/4b32f6fc-cbce-40cd-908f-50564e6534da">
<img width="668" alt="Screenshot 2023-12-22 at 2 01 01 PM" src="https://github.com/woocommerce/woocommerce/assets/1281828/c7b045c2-4680-40ae-8cba-8707a132e7da">
<img width="489" alt="Screenshot 2023-12-22 at 2 00 38 PM" src="https://github.com/woocommerce/woocommerce/assets/1281828/dbdcd1c1-4985-4266-84e7-639453694c8f">

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

1. In block editor insert collection filters block, this will auto insert the rating, attribute and stock filter blocks
2. for each of those blocks, open up the toolbar menu and choose a text color. Note that in the editor the color of the checkbox list text matches the chosen color. Now in the block settings change to dropdown mode. Also see the color reflected in the placeholder text of the dropdown
3. Now save the content and view it on the frontend. See that when opening dropdowns the placeholder and option text color reflect the one chosen.
4. Also go back to the list mode and see in the front end that lists display with the chosen color.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment

Add support for text color to experimental filter blocks

</details>

samueljseay committed:
Introduce Collection Rating Filter, CheckboxList interactivity component.

samueljseay committed:
Revert reintroduction of webpack entry that was removed.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into add/interactivity-rating-filter

samueljseay committed:
Revert accidental changes to AssetsController.

samueljseay committed:
Re-add script queuing for interactivity checkbox and dropdown

samueljseay committed:
Revert accidental change to registered block type.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into add/interactivity-rating-filter

samueljseay committed:
Escape interactive directive namespace attributes in Checkbox/Dropdown.

samueljseay committed:
Remove unused showFilterButton option.

samueljseay committed:
Remove unused elements from block editor.

samueljseay committed:
Consolidate the rating filter JS block into edit.tsx

samueljseay committed:
Refactor to add inspector controls in components/inspector per convention.

ObliviousHarmony committed:
Added Blocks & Admin Aliases

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into add/interactivity-rating-filter

samueljseay committed:
Merge remote-tracking branch 'origin/add/watch-project-aliases' into add/interactivity-rating-filter

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into add/interactivity-rating-filter

samueljseay committed:
Just trying to get build working.

samueljseay committed:
Fix the cp command for blocks so that the contents of build is copied to assets/client/blocks

samueljseay committed:
Remove not strictly needed trailing slash.

samueljseay committed:
Just rm the folder instead

samueljseay committed:
Merge remote-tracking branch 'origin/fix/blocks-build' into add/interactivity-rating-filter

samueljseay committed:
Add collection rating filter to collection template.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into add/interactivity-rating-filter

samueljseay committed:
Revert linter changes to PHP file.

samueljseay committed:
WIP support multi select in interactivity dropdown.

samueljseay committed:
WIP multiple selected items in the dropdown.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/dropdown-multi-select

samueljseay committed:
Move to context for the dropdown and remove state.

samueljseay committed:
Fix styling issues with the multi select mode.

DAreRodz committed:
Refactor Dropdown to start using `wc-each`

DAreRodz committed:
Parse template content inside toVdom

DAreRodz committed:
Implement `data-wc-each` and related

DAreRodz committed:
Refactor Dropdown output

DAreRodz committed:
Expose `getEvaluate` from `hooks.tsx`

DAreRodz committed:
Allow using paths in `wc-each-key`

DAreRodz committed:
Remove unnecessary `wc-each`

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/dropdown-multi-select

samueljseay committed:
Remove a bug in finding index of selected item.

samueljseay committed:
Disable no-op useEffect in directives, add placeholder for multiple select.

samueljseay committed:
Clean up unneeded or fallback.

samueljseay committed:
Merge remote-tracking branch 'davidfork/dev/interactivity-api-directive-each-key' into dev/dropdown-multi-select

samueljseay committed:
Clean up inspector options, properly render multiple or single based on block attrs.

samueljseay committed:
Fix outdated text domains.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

DAreRodz committed:
Do not replace signals during merge

DAreRodz committed:
Clone directive context's value

samueljseay committed:
Remove accidentally duplicated style import.

samueljseay committed:
We can remove force from navigate now that bugs are fixed.

samueljseay committed:
Fix incorrect indentation of comment.

samueljseay committed:
Reintroduce getUrl functionality that consolidates query filter logic.

samueljseay committed:
Add color support to block.json, allow text color passed to Dropdown.php

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/dropdown-support-text-color

samueljseay committed:
Add block level text color rendering support to the editor for relevant filter blocks.

samueljseay committed:
Pass text color down to Dropdown in other filter blocks on PHP side.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

samueljseay committed:
Add background to regex ignore.

samueljseay committed:
Revert autoformatting of price filter block json.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/dropdown-support-text-color

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/collection-filters/inner-blocks/attribute-filter/components/attribute-dropdown.tsx:
```diff
@@ -1,25 +1,38 @@
+};
+
+export const AttributeDropdown = ( { label, textColor }: Props ) => {
+	const StyledFormTokenField = textColor
+		? styled( FormTokenField )`
+				.components-form-token-field__input::placeholder {
+					color: ${ textColor } !important;
+				}
+		  `
+		: FormTokenField;
```
I think this is a good workaround but also a signal to rewrite the dropdown component to remove the `FormTokenField` component.

samueljseay replied:
@dinhtungdu absolutely. I think we need to do that work at some point (after MVP?)

dinhtungdu commented on plugins/woocommerce/src/Blocks/InteractivityComponents/Dropdown.php:
```diff
@@ -90,7 +111,24 @@ class="components-button components-form-token-field__remove-token has-icon"
 								<?php } ?>
 							<?php } ?>
-							<input id="components-form-token-input-1" type="text" autocomplete="off" data-wc-bind--placeholder="state.placeholderText" class="components-form-token-field__input" role="combobox" aria-expanded="false" aria-autocomplete="list" aria-describedby="components-form-token-suggestions-howto-1" value="" data-wc-key="input">
+							<style>
+								#<?php echo esc_html( $input_id ); ?>::placeholder {
+									<?php echo esc_attr( $text_color_style ); ?>
+								}
+							</style>
```
I think we can use [print_late_styles](https://developer.wordpress.org/reference/functions/print_late_styles/) to avoid doing this. The ID need to be consistent between page loads though.

samueljseay replied:
@dinhtungdu I don't think I need to use `print_late_styles` if I just use: [wp_add_inline_style](https://developer.wordpress.org/reference/functions/wp_add_inline_style/) Does that sound right to you?

dinhtungdu replied:
@samueljseay  Ah yeah, you're right! Blocks are rendered before `wp_head` so it's not late style

samueljseay replied:
@dinhtungdu This was a great suggestion btw, thank you. Added in https://github.com/woocommerce/woocommerce/pull/43068/commits/b6b900fd94ea05216d93293ee5125ea0b2641861

dinhtungdu reviewed (commented):
Tested and this is working as intended. I left some review comments. Otherwise, this is LGTM.

samueljseay committed:
Use wp_add_inline_style instead of rendering a style tag to style text color of dropdown placeholder.

samueljseay committed:
[Experimental] Render headings as inner blocks of collection filters (#43109)

* Support render of heading inner block for filters, add inner blocks in Collection Filter template.
* When there are no product ratings don't render rating filter.

samueljseay commented:
@dinhtungdu just noting that I merged the other PR into this, but I've addressed your comment about the stylesheet and this should be ready for another review. 🙇🏻

dinhtungdu commented on plugins/woocommerce/src/Blocks/InteractivityComponents/Dropdown.php:
```diff
@@ -34,25 +32,17 @@ public static function render( $props ) {
-		$wrapper_class = 'multiple' === $select_type ? '' : 'single-selection';
-
-		$input_id = wp_unique_id( 'wc-interactivity-dropdown-input-' );
+		wp_add_inline_style(
+			'wc-interactivity-dropdown',
+			"#$input_id::placeholder {
+					$text_color_style
+			}"
+		);
```
I didn't spot this in the first review, but in my test, we can style the placeholder using CSS only:

```
#wc-interactivity-dropdown-input-3,
#wc-interactivity-dropdown-input-3::placeholder {
	color: inherit;
}
```

samueljseay replied:
Oh good point. I didnt think of that at the time. I will fix this.

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/collection-filters/inner-blocks/attribute-filter/components/attribute-dropdown.tsx:
```diff
@@ -1,25 +1,38 @@
+};
+
+export const AttributeDropdown = ( { label, textColor }: Props ) => {
+	const StyledFormTokenField = textColor
+		? styled( FormTokenField )`
+				.components-form-token-field__input::placeholder {
+					color: ${ textColor } !important;
+				}
+		  `
+		: FormTokenField;
```
It also means that we can use CSS only here also, we don't need to use styled component, hence reducing lots of logic.

dinhtungdu reviewed (approved):
After discussion, I agree to merge this PR at the current state and work on new Dropdown component as the followup.

samueljseay committed:
Fix lint issue.
